### PR TITLE
[1.8] htmlhelp: convert hex escaping to decimal escaping in .hhc/.hhk files

### PR DESCRIPTION
**Pull request by ghost**
### Feature or Bugfix

- Bugfix

### Purpose
.hhc/.hhk files don't recognize hex escaping, we need convert hex escaping to decimal escaping.
For example: `&#x27;` -> `&#39;`

### Detail
If hex escaping (e.g. `&#x27;`) appears in .hhc/.hhk files, chm compiler will simply ignore it.

For example: use `&#x45;` to represent upper letter `E` in .hhc/.hhk files, this letter doesn't apper in chm content/index. After using its decimal form `&#69;` it appears as expected.

Sadly, Python 3's `html.escape(s, quote=True)` function generates a hex escaping `&#x27;` for single quote `'`. This leading to a compiling issue for Python chm document, please see: https://bugs.python.org/issue35482

### Relates
Issue: https://github.com/sphinx-doc/sphinx/issues/5846
Environment: Microsoft HTML Help Compiler: 4.74.8702

### CODE REVIEW

File: sphinx/builders/htmlhelp.py [hunk 1]
```diff
@@ -168,6 +169,18 @@
     'zh_TW': (0x404, 'cp950'),
 }
```
Comment (tk0miya) [inline sphinx/builders/htmlhelp.py:171]: Could you add 2 blank lines before a function and classes?
flake8 warns here and before `HTMLHelpBuilder` class.
https://travis-ci.org/sphinx-doc/sphinx/jobs/471503849#L520-L521

File: sphinx/builders/htmlhelp.py [hunk 2]
```diff
@@ -168,6 +169,18 @@
...
 
+# chm_htmlescape() is a wrapper of htmlescape()
+# .hhc/.hhk files don't recognize hex escaping, we need convert
+# hex escaping to decimal escaping. for example: `&#x27;` -> `&#39;`
+# htmlescape() may generates a hex escaping `&#x27;` for single
+# quote `'`, this wrapper fixes this.
```
Comment (tk0miya) [inline sphinx/builders/htmlhelp.py]: Please move this to a docstring of `chm_htmlescape()`.

File: sphinx/builders/htmlhelp.py [hunk 3]
```diff
@@ -168,6 +169,18 @@
...
+def chm_htmlescape(*args, **kwargs):
+    def convert(matchobj):
```
Comment (tk0miya) [inline sphinx/builders/htmlhelp.py:182]: Please add annotations to these functions.

**Review by tk0miya — CHANGES_REQUESTED** [review]
Looks good. It would be nice if there is testcases. Could you try it please?

**Comment by animalize** [comment]
Now it's prepared for review.
Sorry for my unskilled.

**Comment by ghost** [comment]
Now it's green, sorry for my unskilled.

File: tests/roots/test-root/markup.txt [hunk 4]
```diff
@@ -379,6 +379,10 @@ Index markup
    !Main, !Other
    !single: entry; pair
 
+.. index:: triple-quoted string, Unicode Consortium, raw string
+   single: """; string literal
+   single: '''; https://github.com/sphinx-doc/sphinx/issues/5846
```
Comment (tk0miya) [inline tests/roots/test-root/markup.txt]: Oh, sorry. Please add new test example to `tests/roots/test-[new-test-directory]`. It is commonly used testcase. So I'd like to keep as is.

In addition, you need to add testcase in `tests/test_build_htmlhelp.py`.

Comment (ghost) [reply tests/roots/test-root/markup.txt]: 1.8 branch doesn't have `tests/test_build_htmlhelp.py`.
Even in 2.0 branch, [`tests/test_build_htmlhelp.py`](https://github.com/sphinx-doc/sphinx/blob/master/tests/test_build_htmlhelp.py) is amlost empty.

I don't know how to create a `htmlhelp` testcase...

I replace installed `sphinx/builders/htmlhelp.py ` with this version, then I can build a good `python372rc1.chm`, can this be seem as a test?

Comment (tk0miya) [reply tests/roots/test-root/markup.txt]: Yes, it is manual testing. But we need continuous testing via test code.

Is it hard to check generated `.chm` file? This PR changes the escaping algorithm. It should be confirmed in the case.

Comment (ghost) [reply tests/roots/test-root/markup.txt]: Testcase added.

**Review by tk0miya — APPROVED** [review]
LGTM!

**Comment by tk0miya** [comment]
Merged. Thank you for your work!

File: sphinx/builders/htmlhelp.py [hunk 5]
```diff
@@ -169,6 +170,24 @@
...
+    .hhc/.hhk files don't recognize hex escaping, we need convert
+    hex escaping to decimal escaping. for example: `&#x27;` -> `&#39;`
+    htmlescape() may generates a hex escaping `&#x27;` for single
+    quote `'`, this wrapper fixes this.
+    """
+    def convert(matchobj):
+        # type: (Match[unicode]) -> unicode
+        codepoint = int(matchobj.group(1), 16)
+        return '&#%d;' % codepoint
+    return re.sub(r'&#[xX]([0-9a-fA-F]+);',
```
Comment (tk0miya) [inline sphinx/builders/htmlhelp.py:186]: I noticed `html.escape()` only generates `&#x27;`. Other characters (including unicode) are not changed.
Is this regexp correct? Is it better to use `str.replace()` instead like this?
```
def chm_htmlescape(s):
    s = htmlescape(s)
    s = s.replace('&#x27;', '&#27;')  # re-escape as decimal
    return s
```

Comment (ghost) [reply sphinx/builders/htmlhelp.py:186]: Yes, that works, but I want to sovle this problem completely.

If a document use `&#x45;` to represent upper letter `E`, `&#x45;` will appear in .hhc/.hhk files, so `E` doesn't display in chm's content/index. After using its decimal form `&#69;` it displays as expected. Though this kind of document is very rare.

Comment (ghost) [reply sphinx/builders/htmlhelp.py:186]: Maybe you are right, `&#x45;` will be escaped to `&amp;#x45;` by the `Writer`, so we don't need worry about it...

Comment (tk0miya) [reply sphinx/builders/htmlhelp.py:186]: I found this is current behavior:
```
def test_chm_htmlescape():
    assert chm_htmlescape('Hello world') == 'Hello world'
    assert chm_htmlescape(u'Unicode 文字') == u'Unicode 文字'

    if PY2:
        assert chm_htmlescape('<Hello> "world"') == '&lt;Hello&gt; "world"'
        assert chm_htmlescape('<Hello> "world"', True) == '&lt;Hello&gt; &quot;world&quot;'
        assert chm_htmlescape('<Hello> "world"', False) == '&lt;Hello&gt; "world"'
    else:
        assert chm_htmlescape('<Hello> "world"') == '&lt;Hello&gt; &quot;world&quot;'
        assert chm_htmlescape('<Hello> "world"', True) == '&lt;Hello&gt; &quot;world&quot;'
        assert chm_htmlescape('<Hello> "world"', False) == '&lt;Hello&gt; "world"'

    if PY2:
        assert chm_htmlescape("Hello 'world'") == "Hello 'world'"
        assert chm_htmlescape("Hello 'world'", True) == "Hello 'world'"
        assert chm_htmlescape("Hello 'world'", False) == "Hello 'world'"
    else:
        assert chm_htmlescape("Hello 'world'") == "Hello &#39;world&#39;"
        assert chm_htmlescape("Hello 'world'", True) == "Hello &#39;world&#39;"
        assert chm_htmlescape("Hello 'world'", False) == "Hello 'world'"
```

It seems single quote does not escaped on py2. Is this intended? If none, I'll fix it also from now on.

Comment (ghost) [reply sphinx/builders/htmlhelp.py:186]: `htmlescape()` is defined in file `sphinx.util.pycompat.py`:
```python
# htmlescape()
if PY3:
    from html import escape as htmlescape
else:
    from cgi import escape as htmlescape  # NOQA
```

In py3, we use `html.escape()` from stdlib, it convert signle quote if `quote=True`.
https://github.com/python/cpython/blob/4a9ee26750aa8cb37b5072b2bb4dd328819febb4/Lib/html/__init__.py#L24
This function escapes single quote.

In py2, we use `cgi.escape()` from stdlib:
https://github.com/python/cpython/blob/bc6f74a520112d25ef40324e3de4e8187ff2835d/Lib/cgi.py#L1063-L1072
This function doesn't escape single quote.

I think it's not a matter if we always use `"` to wrap attributes in .html templates.

Comment (tk0miya) [reply sphinx/builders/htmlhelp.py:186]: Okay, I'll keep this as is.

**Pull request merged**